Make `reflection.klass` raise if `polymorphic?` not to be misused

kamipo opened the pull request:
This is an alternative of #31877 to fix #31876 caused by #28808.

This issue was caused by a combination of several loose implementation.

* finding automatic inverse association of polymorphic without context (caused by #28808)
* returning `klass` even if `polymorphic?` (exists before #28808)
* loose verification by `valid_inverse_reflection?` (exists before #28808)

This makes `klass` raise if `polymorphic?` not to be misused.
This issue will not happen unless polymorphic `klass` is misused.

Fixes #31876.
Closes #31877.

kamipo committed:
Make `reflection.klass` raise if `polymorphic?` not to be misused

This is an alternative of #31877 to fix #31876 caused by #28808.

This issue was caused by a combination of several loose implementation.

* finding automatic inverse association of polymorphic without context (caused by #28808)
* returning `klass` even if `polymorphic?` (exists before #28808)
* loose verification by `valid_inverse_reflection?` (exists before #28808)

This makes `klass` raise if `polymorphic?` not to be misused.
This issue will not happen unless polymorphic `klass` is misused.

Fixes #31876.
Closes #31877.

matthewd commented on activerecord/lib/active_record/reflection.rb:
```diff
@@ -416,6 +416,9 @@ def mapping
     class AssociationReflection < MacroReflection #:nodoc:
       def compute_class(name)
+        if polymorphic?
+          raise NameError, "Polymorphic association does not support to compute class."
```
Hmm.. I see the advantage of leaning on the existing code, but I think making this a NameError is still a bit confusing: normally a NameError here is the fault of the data (a bad/unknown class name), whereas in this case it's the fault of the calling code (breaking a contract by calling a method when it shouldn't). Better to keep them separate, I think -- and then pull in the conditional from #31877 to ensure we conform to that contract.

(Otherwise some other code might also rescue NameError, and do something else -- and then they would miss out on the warning that they're misusing the API.)

kamipo replied:
Your requested changes are

* Do not use `NameError` (use `ArgumentError` instead, or another?)
* Do not attempt to automatically find the inverse of a polymorphic not to attempt to call `klass` of polymorphic

Right?

kamipo committed:
Do not attempt to automatically find the inverse of a polymorphic `belongs_to` association

We can't automatically find the inverse of a polymorphic `belongs_to`
association without context.

[Ryuta Kamizono & Eric K Idema]

rafaelfranca commented:
What misused means here? Is https://github.com/rails/rails/issues/34312 a misuse of this feature? At first glance I see nothing wrong in that code

kamipo commented:
At a glance I agree that nothing wrong in that code, it seems there is a bug around polymorphic through association, so I'll take a closer look at that in this weekend.

BrentW commented:
I'm wondering if you've had a chance to look at related issue https://github.com/rails/rails/issues/34312
